internal/domain/release.go

```diff
@@ -423,58 +425,72 @@ func (r *Release) downloadTorrentFile(ctx context.Context) error {
+			return retry.Unrecoverable(errors.New("unrecoverable error downloading torrent (%v) file (%v) - status code: %d - check indexer keys for %s", r.TorrentName, r.TorrentURL, resp.StatusCode, r.Indexer))
 
 		case http.StatusMethodNotAllowed:
 			return retry.Unrecoverable(errors.New("unrecoverable error downloading torrent (%v) file (%v) from '%v' - status code: %d. Check if the request method is correct", r.TorrentName, r.TorrentURL, r.Indexer, resp.StatusCode))
 
 		case http.StatusNotFound:
 			return errors.New("torrent %s not found on %s (%d) - retrying", r.TorrentName, r.Indexer, resp.StatusCode)
 
-		case http.StatusInternalServerError, http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout:
+		case http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout:
 			return errors.New("server error (%d) encountered while downloading torrent (%v) file (%v) from '%v' - retrying", resp.StatusCode, r.TorrentName, r.TorrentURL, r.Indexer)
 
+		case http.StatusInternalServerError:
+			return errors.New("server error (%d) encountered while downloading torrent (%v) file (%v) - check indexer keys for %s", resp.StatusCode, r.TorrentName, r.TorrentURL, r.Indexer)
+
 		default:
 			return retry.Unrecoverable(errors.New("unexpected status code %d: check indexer keys for %s", resp.StatusCode, r.Indexer))
 		}
 
-		// Check if the Content-Type header is correct
-		contentType := resp.Header.Get("Content-Type")
-
-		if strings.Contains(contentType, "text/html") {
-			return retry.Unrecoverable(errors.New("unexpected content type '%s': check indexer keys for %s", contentType, r.Indexer))
-		}
-
 		resetTmpFile := func() {
 			tmpFile.Seek(0, io.SeekStart)
 			tmpFile.Truncate(0)
 		}
 
-		// Write the body to file
-		if _, err := io.Copy(tmpFile, resp.Body); err != nil {
-			resetTmpFile()
-			return errors.Wrap(err, "error writing downloaded file: %v", tmpFile.Name())
+		// Read the body into bytes
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return errors.Wrap(err, "error reading response body")
 		}
 
-		meta, err := metainfo.LoadFromFile(tmpFile.Name())
+		// Create a new reader for bodyBytes
+		bodyReader := bytes.NewReader(bodyBytes)
+
+		// Try to decode as torrent file
+		meta, err := metainfo.Load(bodyReader)
 		if err != nil {
 			resetTmpFile()
-			return errors.Wrap(err, "metainfo could not load file contents: %v", tmpFile.Name())
+
+			// explicitly check for unexpected content type that match html
+			var bse *bencode.SyntaxError
+			if errors.As(err, &bse) {
+				// regular error so we can retry if we receive html first run
+				return errors.Wrap(err, "metainfo unexpected content type, got HTML expected a bencoded torrent. check indexer keys for %s - %s", r.Indexer, r.TorrentName)
+			}
+
+			return retry.Unrecoverable(errors.Wrap(err, "metainfo unexpected content type. check indexer keys for %s - %s", r.Indexer, r.TorrentName))
```

Comment:
is this really fatal? shouldn't we retry.

Reply:
The bencode.SyntaxError is what happens when it tries to parse `<html>..</html>` and that is handled above without unrecoverable so it can retry if for some reason it gets different content first time because it's too fast or something.

The last err is for all other cases, and there is a lot of other errors metainfo can return but all of them are bad. `Metainfo` also does a horrendous amount of panics internally 👀

Reply:
I wrote this eons ago, maybe it's time for it's debut.

```
func Validate(data string) error {
	depth := 0
	buf := strings.NewReader(data)
	for {
		b, err := buf.ReadByte()
		if err == io.EOF {
			break
		}

		switch b {
		case 'd', 'l':
			depth++
		case 'e':
			depth--
			if depth < 0 {
				return fmt.Errorf("Unexpected depth at %d: %c", depth, b)
			}
		case 'i':
			init := true
			for {
				i, err := buf.ReadByte()
				if err == io.EOF && depth != 0 {
					return fmt.Errorf("Unexpected end to integerstream at a depth of: %d", depth)
				}

				if init {
					init = false
					if i == '-' {
						if k, err := buf.ReadByte(); k == '0' || err != nil {
							return fmt.Errorf("Unexpected value in integerstream at a depth of: %d| %d", depth, k)
						}

						buf.UnreadByte()
						continue
					}
				}

				if i < '0' || i > '9' {
					if i == 'e' {
						break
					}

					return fmt.Errorf("Unexpected end to integerstream at a depth of: %d | %c", depth, i)
				}
			}
		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
			i := int64(b - '0')
			for {
				ru, err := buf.ReadByte()
				if err == io.EOF && depth != 0 {
					return fmt.Errorf("Unexpected end to bytestream at a depth of: %d", depth)
				}

				k := int64(ru - '0')
				if k < 0 || k > 9 {
					if ru == ':' {
						break
					}

					return fmt.Errorf("Unexpected length in bytestream at a depth of: %d | %c", depth, ru)
				}

				i *= 10
				i += k
			}

			buf.Seek(i, io.SeekCurrent)

		default:
			return fmt.Errorf("Unexpected character in bytestream at a depth of: %d | %c", depth, b)
		}
	}

	if depth != 0 {
		return fmt.Errorf("Unexpected depth at end of stream: %d", depth)
	}

	return nil
}
```